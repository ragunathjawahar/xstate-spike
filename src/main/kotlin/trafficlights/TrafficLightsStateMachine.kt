package trafficlights

import io.redgreen.kstate.annotation.EffectEvent
import io.redgreen.kstate.annotation.Initial
import io.redgreen.kstate.annotation.Next
import io.redgreen.kstate.annotation.On
import io.redgreen.kstate.annotation.StateMachine
import io.redgreen.kstate.contract.Effect
import io.redgreen.kstate.contract.Event
import io.redgreen.kstate.contract.State
import trafficlights.TrafficLightsEffect.BeginCountDown
import trafficlights.TrafficLightsEvent.CountDownElapsed
import trafficlights.TrafficLightsState.Green

@StateMachine
@Initial(Green::class, effects = [BeginCountDown::class])
class TrafficLightsStateMachine

sealed class TrafficLightsState : State {
  @On(
    CountDownElapsed::class,
    Next(Green::class, effects = [BeginCountDown::class])
  )
  object Red : TrafficLightsState() {
    override fun toString(): String = this::class.java.simpleName
  }

  @On(
    CountDownElapsed::class,
    Next(Red::class, effects = [BeginCountDown::class])
  )
  object Yellow : TrafficLightsState() {
    override fun toString(): String = this::class.java.simpleName
  }

  @On(
    CountDownElapsed::class,
    Next(Yellow::class, effects = [BeginCountDown::class])
  )
  object Green : TrafficLightsState() {
    override fun toString(): String = this::class.java.simpleName
  }
}

sealed class TrafficLightsEvent : Event {
  @EffectEvent(BeginCountDown::class)
  object CountDownElapsed : TrafficLightsEvent()
}

sealed class TrafficLightsEffect : Effect {
  object BeginCountDown : TrafficLightsEffect()
}
