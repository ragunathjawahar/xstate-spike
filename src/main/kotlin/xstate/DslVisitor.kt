package xstate

import kotlin.reflect.KClass
import xstate.visitors.mobius.Reducer

interface DslVisitor {
  fun onName(name: String)

  fun onInitialState(
    initialState: KClass<out Any>,
    effects: Set<KClass<out Any>>
  )

  fun onState(
    state: KClass<out Any>,
    final: Boolean
  )

  fun onTransition(
    event: KClass<out Any>,
    next: KClass<out Any>,
    effects: Set<KClass<out Any>>,
    reducer: KClass<out Reducer<out Any, out Any>>
  )
}
