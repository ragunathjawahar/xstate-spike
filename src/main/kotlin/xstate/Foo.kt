package xstate

import kotlin.LazyThreadSafetyMode.NONE

fun <T> fastLazy(initializer: () -> T): Lazy<T> =
  lazy(NONE, initializer)

fun <S : Any, E : Any, F : Any, V: DslVisitor> runVisitor(
  machine: Machine<S, E, F>,
  dslVisitor: V
): V =
  machine.visit(dslVisitor)
