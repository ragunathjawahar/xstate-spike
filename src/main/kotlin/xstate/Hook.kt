package xstate

import kotlin.reflect.KClass

interface Hook {
  fun onName(name: String)
  fun onInitialState(initialState: KClass<out Any>)
  fun onState(state: KClass<out Any>)
  fun onTransition(event: KClass<out Any>, next: KClass<out Any>)
}
