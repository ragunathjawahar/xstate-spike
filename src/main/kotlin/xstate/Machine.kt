package xstate

import kotlin.reflect.KClass

class Machine<S : Any, E : Any, F : Any> private constructor(
  private val dslVisitor: DslVisitor,
  private val name: String,
  private val initialState: KClass<out S>, // TODO Move this into the DSL
) {
  companion object {
    fun <S : Any, E : Any, F : Any> create(
      dslVisitor: DslVisitor,
      name: String,
      initialState: KClass<out S>, // TODO Move this into the DSL
      definition: Machine<S, E, F>.() -> Unit,
    ): Machine<S, E, F> {
      return Machine<S, E, F>(dslVisitor, name, initialState)
        .apply { definition(this) }
    }
  }

  fun states(
    builder: States<S, E, F>.() -> Unit
  ) {
    dslVisitor.onName(name)
    dslVisitor.onInitialState(initialState)
    builder(States(dslVisitor))
  }
}

class States<S : Any, E : Any, F : Any>(
  private val dslVisitor: DslVisitor
) {
  fun state(
    state: KClass<out S>,
    definition: State<S, E, F>.() -> Unit
  ) {
    dslVisitor.onState(state)
    definition(State(dslVisitor))
  }
}

class State<S : Any, E : Any, F : Any>(
  private val dslVisitor: DslVisitor
) {
  fun on(
    event: KClass<out E>,
    next: KClass<out S>,
    effects: Set<KClass<out F>> = emptySet()
  ) {
    dslVisitor.onTransition(event, next, effects)
  }
}
