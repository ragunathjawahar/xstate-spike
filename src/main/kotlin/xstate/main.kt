package xstate

import trafficlights.TrafficLightsEffect
import trafficlights.TrafficLightsEvent
import trafficlights.TrafficLightsEvent.CountDownElapsed
import trafficlights.TrafficLightsState
import trafficlights.TrafficLightsState.Green
import trafficlights.TrafficLightsState.Red
import trafficlights.TrafficLightsState.Yellow
import xstate.visitors.MobiusVisitor

fun main() {
  val visitor = MobiusVisitor()

  Machine<TrafficLightsState, TrafficLightsEvent, TrafficLightsEffect>(visitor, "Traffic Lights", Green::class) {
    states {
      state(Green::class) {
        on(CountDownElapsed::class, next = Yellow::class)
      }
      state(Yellow::class) {
        on(CountDownElapsed::class, next = Red::class)
      }
      state(Red::class) {
        on(CountDownElapsed::class, next = Green::class)
      }
    }
  }()

  println("Initial state:" + visitor.initialState)
  println("Green + CountDownElapsed = " + visitor.updateFunction.invoke(Green, CountDownElapsed))
  println("Yellow + CountDownElapsed = " + visitor.updateFunction.invoke(Yellow, CountDownElapsed))
  println("Red + CountDownElapsed = " + visitor.updateFunction.invoke(Red, CountDownElapsed))
}
