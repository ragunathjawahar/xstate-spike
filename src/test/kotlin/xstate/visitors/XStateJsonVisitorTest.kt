package xstate.visitors

import org.approvaltests.Approvals
import org.junit.jupiter.api.Test
import xstate.examples.trafficlights.trafficLightsStateMachine
import xstate.runVisitor

internal class XStateJsonVisitorTest {
  @Test
  fun `traffic lights`() {
    Approvals
      .verify(runVisitor(trafficLightsStateMachine, XStateJsonVisitor()).json)
  }
}
