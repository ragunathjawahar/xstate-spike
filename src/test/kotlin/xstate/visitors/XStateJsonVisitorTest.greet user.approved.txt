{
  "id": "Open Project",
  "initial": "GettingGitUsername",
  "states": {
    "GettingGitUsername": {
      "on": {
        "GitUsernameFound": "PersonalizedGreeting",
        "GitUsernameNotFound": "GenericGreeting",
        "UnableToGetGitUsername": "GenericGreeting"
      }
    },
    "PersonalizedGreeting": {
      "type": "final"
    },
    "GenericGreeting": {
      "type": "final"
    }
  }
}

~ NOTE: Link to the viz tool: https://xstate.js.org/viz/ ~